### PR TITLE
[Datasets] Fix boolean tensor column slicing.

### LINKED ISSUE #20825: [Bug] TensorDType conversion causes boolean tensor columns to duplicate across batches
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Others

### What happened + What you expected to happen

When using the `TensorDType` to cast a binary valued column in a dataset to a tensor type, it breaks the `iter_batches` behavior by causing the first batch of data (only for the binary column) to repeat across every subsequent batch.

Binary columns should behave the same as any other column, and properly use new data for subsequent batches.

### Versions / Dependencies

Repro in 1.7.0 and 1.8.0.

### Reproduction script

```
df = pd.DataFrame(data=[[False], [True], [False], [False]], columns=['target'])
ds = ray.data.from_pandas(df)
pipe = ds.repeat(1).iter_datasets()

columns = ['target']
def to_tensors(df: pd.DataFrame) -> pd.DataFrame:
    for c in columns:
        df[c] = df[c].astype(TensorDtype())
    return df

dataset = next(pipe)
for batch in dataset.map_batches(
    to_tensors, batch_format="pandas"
).iter_batches(
    prefetch_blocks=0,
    batch_size=2,
    batch_format="pandas"
):
    print(batch)
```

Note the output is:

```
target
0  False
1   True
  target
0  False
1   True
```

It should of course be:

```
target
0  False
1   True
  target
0  False
1   False
```

If you remove the `map_batches` call, it will produce the correct results.

### Anything else

cc @clarkzinzow 

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by clarkzinzow**
Arrow byte-packs boolean arrays, with 8 entries per byte, and both Arrow and NumPy utilize bit-based offsets for indexing into data buffers. This PR adds support for properly indexing into boolean tensor columns by using bit-based offsets for such columns.

## Related issue number

Closes #20825 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/tests/test_dataset.py [hunk 1]
```diff
@@ -522,6 +524,23 @@ def test_arrow_tensor_array_getitem(ray_start_regular_shared):
         np.testing.assert_array_equal(t_arr2[idx - 1], arr[idx])
 
 
+@pytest.mark.parametrize("test_arr", [
```
Comment (scv119) [inline python/ray/data/tests/test_dataset.py:527]: does it makes sense to test int32_t and int64_t for completeness?

**Review by scv119 — APPROVED** [review]

File: python/ray/data/extensions/tensor_extension.py [hunk 2]
```diff
@@ -1282,14 +1282,21 @@ def _to_numpy(self,
...
-        item_byte_width = storage_list_type.value_type.bit_width // 8
+        value_type = storage_list_type.value_type
+        if pa.types.is_boolean(value_type):
+            # Boolean array buffers are byte-packed, with 8 entries per byte,
+            # and are accessed via bit offsets.
+            buffer_item_width = value_type.bit_width
+        else:
+            # We assume all other array types are accessed via byte array
+            # offsets.
+            buffer_item_width = value_type.bit_width // 8
```
Comment (ericl) [inline python/ray/data/extensions/tensor_extension.py:1293]: Could we get this magic number "8" programmatically from Arrow instead of special casing?

Comment (clarkzinzow) [reply python/ray/data/extensions/tensor_extension.py:1293]: I tried to find a way to do that, but this is fundamentally a value type question: boolean arrays are byte-packed and use bit offsets (see [Pandas writer](https://github.com/apache/arrow/blob/415439cd0a30732830249c6924dff945b88630b4/cpp/src/arrow/python/arrow_to_pandas.cc#L1317-L1327) and [indexing logic](https://github.com/apache/arrow/blob/e903a214525dd6dcd8e57b20958a65dc678be47d/cpp/src/arrow/array/array_primitive.h#L50-L53)), all other primitive type arrays aren't byte-packed and [use byte offsets](https://github.com/apache/arrow/blob/415439cd0a30732830249c6924dff945b88630b4/cpp/src/arrow/python/arrow_to_pandas.cc#L282-L284). I haven't found a way to reliably determine whether a type's buffer is byte-addressable or bit-addressable.

Comment (clarkzinzow) [reply python/ray/data/extensions/tensor_extension.py:1293]: The magic number "8" is just `CHAR_BIT`, the number of bits in a `char` byte, which is 8 on all platforms that we wish to support (`CHAR_BIT == 8` on all POSIX systems and Windows). I couldn't find a way to programmatically retrieve `CHAR_BIT` from Python, the closest I could find was `os.sysconf("SC_CHAR_BIT")`, but that's a GNU extension, not available on all non-Unix platforms nor even all Unix platforms.

I can also define a module-level `NUM_BITS_IN_BYTE = 8` constant if you feel that this magic number is too opaque?

Comment (ericl) [reply python/ray/data/extensions/tensor_extension.py:1293]: > all other primitive type arrays aren't byte-packed and use byte offsets.

Thanks, this was what I wanted to know.

**Pull request merged**